unified-scheduler-pool/src/lib.rs

```diff
@@ -530,9 +530,68 @@ impl<S: SpawnableScheduler<TH>, TH: TaskHandler> ThreadManager<S, TH> {
     }
 
     fn start_threads(&mut self, context: &SchedulingContext) {
-        let (mut runnable_task_sender, runnable_task_receiver) =
+        // Firstly, setup bi-directional messaging between the scheduler and handlers to pass
+        // around tasks, by creating 2 channels (one for to-be-handled tasks from the scheduler to
+        // the handlers and the other for finished tasks from the handlers to the scheduler).
+        // Furthermore, this pair of channels is duplicated to work as a primitive 2-level priority
+        // queue, totalling 4 channels.
+        //
+        // This quasi-priority-queue arrangement is desired as an optimization to prioritize
+        // blocked tasks.
+        //
+        // As a quick background, SchedulingStateMachine doesn't throttle runnable tasks at all.
+        // Thus, it's likely for to-be-handled tasks to be stalled for extended duration due to
+        // excessive buffering (commonly known as buffer bloat). Normally, this buffering isn't
+        // problematic and actually intentional to fully saturate all the handler threads.
+        //
+        // However, there's one caveat: task dependencies. It can be hinted with tasks being
+        // blocked, that there could be more similarly-blocked tasks in the future. Empirically,
+        // clearing these linearized long runs of blocking tasks out of the buffer is delaying bank
+        // freezing while only using 1 handler thread or two near the end of slot, deteriorating
+        // the overall concurrency.
```

Comment:
Curious about the choice here.

> It can be *hinted* with tasks blocked, that there could be more similarly-blocked tasks in the future.

This seems odd to me to essentially allow queue-skipping if the task itself were blocked because we think it **might** block other tasks, when we have a direct method of determining whether or not there are further blocked tasks behind this one.
Each task has the UsageQueue's for the accounts it needs, each of which can tell us whether or not there are more than 1 blocked task.

Probably not ideal to loop through these, but would definitely be curious to here your thoughts on this hint vs direct.

Reply:
wow. really good question: a4f06a08ce9e0d53bf963a92d0fe288131d60ef3